File: civitasJava/src/civitas/CivitasJuego.java
package civitas;
import java.util.ArrayList;

/**
 * @author Yesenia González Dávila
 * @author Esther García Gallego
 * Grupo B.3
 */

public class CivitasJuego {

    private Tablero tablero;
    private MazoSorpresas mazo;
    private ArrayList<Jugador> jugadores;                       
    private EstadosJuego estado;
    private GestorEstados gestorEstados;

    private int indiceJugadorActual;

    static final int numJugadores = 2;                          
    static final int casillaCarcel = 3;                                      
    static final int numCasillas = 20;               

    /** Constructor. */
    public CivitasJuego (ArrayList<String> nombres){      
        jugadores = new ArrayList<> ();

        for (int i = 0; i < numJugadores; i++)  
            jugadores.add (new Jugador (nombres.get(i)));

        gestorEstados = new GestorEstados();
        estado = gestorEstados.estadoInicial();

        indiceJugadorActual = Dado.getInstance().quienEmpieza (numJugadores);

        tablero = new Tablero(casillaCarcel);   
        mazo = new MazoSorpresas (true);       // Lo inicializamos con el debug activado 
        
        inicializaMazoSorpresas (tablero);
        inicializaTablero (mazo);
    }

    /** Inicializa el tablero. */
    private void inicializaTablero (MazoSorpresas mazo){
        TituloPropiedad t1, t2, t3;

        // Calles
        int cont = 1;
        final int alquiler = 100, hipBase = 50, precioCompra = 150, precioEdif = 200;
        final float factRev = 2.0f;
        final String calle = "Calle ";    

        // Casilla impuesto
        final float cantidad_impuesto = 150f; 
        
        for (int i=1 ; i < numCasillas; i++)       // Carcel se añade automáticamente (en c3), y la salida ya está en 0
            switch (i) {
                case 2:     // Sorpresa 1
                    tablero.añadeCasilla (new CasillaSorpresa ("Sorpresa 1", mazo));
//                    tablero.añadeCasilla ( new Casilla (mazo, "Sorpresa 1"));
                    break;
                    
                case 4:     // Impuesto
                    tablero.añadeCasilla(new CasillaImpuesto ("Impuesto de " + cantidad_impuesto + ".", cantidad_impuesto));
                    break;
                    
                case 7:      // Sorpresa 2
                    tablero.añadeCasilla (new CasillaSorpresa ("Sorpresa 2", mazo));
                    break;
                    
                case 8:      // Juez
                    tablero.añadeJuez();
                    break;
                    
                case 10:     // Sorpresa 3
                    tablero.añadeCasilla (new CasillaSorpresa ("Sorpresa 3", mazo));
                    break;
                    
                case 12:     // Parking
                    tablero.añadeCasilla(new Casilla ("Parking"));
                    break;
                    
                case 3:      // Carcel (se añade automaticamente) --> si no se pone el case entra en default y crea calle
                    break;
                    
                default:    // Calles: casillas 1, 5, 6, 9, 11-19
                    tablero.añadeCasilla (new CasillaCalle
                           (new TituloPropiedad (calle + cont++, alquiler, factRev, hipBase, precioCompra, precioEdif)));
                    break;
            }
    }
    
    /** Inicializa el mazo. */
    private void inicializaMazoSorpresas (Tablero tablero){
        
        int valor = 200;  // <-- Para sorpresas PAGARCOBRAR, PORJUGADOR y PORCASAHOTEL
        int ir_casilla_juez = 8;  
        int ir_a_calle = 9;
        int ir_a_calle2 = 12;
        
        
        // Llevar a calle
        mazo.alMazo (new SorpresaIrCasilla (tablero, ir_a_calle, " ¡Vas a la calle de la casilla " + ir_a_calle +"!"));
        
        // Llevar a juez
        mazo.alMazo (new SorpresaIrCasilla (tablero, ir_casilla_juez, " Irás al juez..."));
        
        // Llevar a calle 2 
        mazo.alMazo (new SorpresaIrCasilla (tablero, ir_a_calle2, " ¡Vas a la calle de la casilla " + ir_a_calle2 +"!"));
        
        // Ir carcel
        mazo.alMazo (new SorpresaIrCarcel (tablero));
        
        // Salir carcel
        mazo.alMazo (new SorpresaSalirCarcel (mazo));
        
        // Por jugador, positiva (recibe) y negativa (paga)
        mazo.alMazo (new SorpresaPorJugador (valor, " ¡Recibes 200 civiMonedas de cada jugador!"));  
        mazo.alMazo (new SorpresaPorJugador (valor*-1, " Tienes que pagarle 200 civiMonedas a cada jugador :( ..."));  
        
        // Por casa hotel, positiva y negativa
        mazo.alMazo (new SorpresaPorCasaHotel (valor, " ¡Recibes 200 civiMonedas por cada casa y hotel que tengas!"));
        mazo.alMazo (new SorpresaPorCasaHotel (valor*-1, " Tienes que pagar 200 civiMonedas por cada casa y hotel que tengas :( ..." ));
        
        // Pagar cobrar, positiva y negativa
        mazo.alMazo (new SorpresaPagarCobrar (valor, " ¡Cobras 200 civiMonedas!" ));
        mazo.alMazo (new SorpresaPagarCobrar (valor*-1, " Tienes que pagar 200 civiMonedas :(..." ));
        
    }

    /** Avanza jugador. */
    private void avanzaJugador() {
        
        // Declaramos al jugador actual y su posicion
        Jugador jugadorActual = getJugadorActual();
        int posicionActual = jugadorActual.getNumCasillaActual();
        
        // Calculamos su nueva posicion tirando el dado
        int tirada = Dado.getInstance().tirar();
        int posicionNueva = tablero.nuevaPosicion (posicionActual, tirada);
        
        // Declaramos la casilla en la que está la nueva posición
        Casilla casilla = tablero.getCasilla(posicionNueva);
        
        // Comprobamos si ha pasado por salida para que el jugador reciba el dinero en tal caso
        contabilizarPasosPorSalida (jugadorActual);
        
        // Movemos al jugador a la nueva posición
        jugadorActual.moverACasilla(posicionNueva);
        
        // Actualizamos la casilla y volvemos a comprobar si ha pasado por salida
        casilla.recibeJugador (indiceJugadorActual, jugadores);
        contabilizarPasosPorSalida (jugadorActual);
    } 
    
    public Jugador getJugadorActual(){
      return jugadores.get(indiceJugadorActual);
    }

    public Casilla getCasillaActual(){   
      return (tablero.getCasilla (getJugadorActual().getNumCasillaActual()));
    }

    private void contabilizarPasosPorSalida(Jugador jugadorActual){
      while (tablero.getPorSalida() > 0)
        jugadorActual.pasaPorSalida();
    }

    private void pasarTurno(){
      indiceJugadorActual = (++indiceJugadorActual) % numJugadores;
    }
    
    public OperacionesJuego siguientePaso() {
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
        OperacionesJuego operacion = gestorEstados.operacionesPermitidas (jugadorActual, estado);
        
        if (operacion == OperacionesJuego.PASAR_TURNO){
            pasarTurno();
            siguientePasoCompletado(operacion);
        }
        else if (operacion == OperacionesJuego.AVANZAR) {
            avanzaJugador();
            siguientePasoCompletado(operacion);
        }
        return operacion;
    }

    public void siguientePasoCompletado (OperacionesJuego operacion){
      estado = gestorEstados.siguienteEstado(getJugadorActual(), estado, operacion);
    }

    public boolean construirCasa (int ip){
       return getJugadorActual().construirCasa(ip);
    }

    public boolean construirHotel (int ip){
      return getJugadorActual().construirHotel(ip);
    }

    public boolean vender (int ip){
      return getJugadorActual().vender(ip);
    }
    
    public boolean comprar () {
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
        Casilla casilla = tablero.getCasilla (jugadorActual.getNumCasillaActual());
        TituloPropiedad titulo = casilla.getTituloPropiedad();
        return (jugadorActual.comprar(titulo));
    }
    
    public boolean hipotecar (int ip){
      return getJugadorActual().hipotecar(ip);
    }

    public boolean cancelarHipoteca (int ip){
      return getJugadorActual().cancelarHipoteca(ip);
    }

    public boolean salirCarcelPagando (){
      return getJugadorActual().salirCarcelPagando();
    }

    public boolean salirCarcelTirando (){
      return getJugadorActual().salirCarcelTirando();
    }

    public boolean finalDelJuego(){
      boolean finaljuego = false;

      for (int i = 0; i < numJugadores && !finaljuego; i++)
          finaljuego = jugadores.get(i).enBancarrota();
      
      return finaljuego;
    }

    //@Override
    public int compareTo(Jugador j){
      return Float.compare(getJugadorActual().getSaldo(), j.getSaldo());
    }

    public ArrayList<Jugador> ranking() {
        Jugador max;
        ArrayList<Jugador> playersrank = new ArrayList();
        
        ArrayList<Jugador> jugadores_aux = (ArrayList)jugadores.clone();
        
        for (int i = 0; i < numJugadores; i++) {
            max = jugadores_aux.get (0);
            for (int j = 1; j < jugadores_aux.size(); j++) 
                if (max.compareTo(jugadores_aux.get(j)) < 0) 
                    max = jugadores_aux.get(j);
            playersrank.add(max);
            jugadores_aux.remove(max);
        }    
        
        return playersrank;
    }

}

File: civitasRuby/pkg/civitasRuby-0.0.1/lib/civitas/civitas_juego.rb
#encoding:utf-8

=begin
Authors: Esther García Gallego
         Yesenia González Dávila
         Grupo B3
=end

require_relative 'jugador.rb'
require_relative 'gestor_estados.rb'
require_relative 'Dado'
require_relative 'Tablero'
require_relative 'MazoSorpresas'
require_relative 'TituloPropiedad'


module Civitas
  class CivitasJuego
    
    @@NumJugadores = 2  
    @@CasillaCarcel = 3
    @@NumCasillas = 20
    
    # Añadimos consultor de numero de jugadores para poder consultarlo en el controlador
    # Consultor de propiedades para acceder a el en vista textual.
    attr_reader :NumJugadores, :propiedades
    
    def initialize(nombres)
      @jugadores = []
      
      for i in 0..@@NumJugadores
        @jugadores.insert(i,Civitas::Jugador::new(nombres[i])) # Crea el jugador
      end
      
      @gestorEstados = Civitas::Gestor_estados.new()
      @estado = @gestorEstados.estado_inicial()
      
      #Dado.instance.set
      @indiceJugadorActual = Dado.instance().quienEmpieza(@@NumJugadores)
      
      @mazo = MazoSorpresas.new(true)
      @tablero = Tablero.new(@@CasillaCarcel)
      inicializaMazoSorpresas(@tablero)
      inicializaTablero(@mazo)
      
    end
           
    def getJugadorActual
      jugador = @jugadores[@indiceJugadorActual]
    end
    
    def getCasillaActual
      jugador = getJugadorActual()
      numCasilla = jugador.numCasillaActual
      @tablero.getCasilla(numCasilla)
    end
    
    def siguientePaso
      jugadorActual = @jugadores[@indiceJugadorActual]
      operacion = @gestorEstados.operaciones_permitidas(jugadorActual, @estado)
      
      if operacion == Civitas::Operaciones_juego::PASAR_TURNO
        pasarTurno()
        siguientePasoCompletado(operacion)
      elsif operacion == Civitas::Operaciones_juego::AVANZAR
        avanzaJugador()
        siguientePasoCompletado(operacion)
      end
      
      operacion
    end
    
    def siguientePasoCompletado(operacion)
      @estado = @gestorEstados.siguiente_estado(@jugadores[@indiceJugadorActual], @estado, operacion)
    end
    
    def construirCasa(ip)
      getJugadorActual().construirCasa(ip)
    end
    
    def construirHotel(ip)
      getJugadorActual().construirHotel(ip)
    end
    
    def vender(ip)
      getJugadorActual().vender(ip)
    end
    
    def comprar
      jugadorActual = @jugadores.at(@indiceJugadorActual)
      casilla =@tablero.getCasilla(jugadorActual.numCasillaActual)
      titulo = casilla.tituloPropiedad()
      jugadorActual.comprar(titulo)
    end
    
    def hipotecar(ip)
      getJugadorActual().hipotecar(ip)
    end
    
    def cancelarHipoteca(ip)
      getJugadorActual().cancelarHipoteca(ip)
    end
    
    def salirCarcelPagando
      getJugadorActual().salirCarcelPagando()
    end
    
    def salirCarcelTirando
      getJugadorActual().salirCarcelTirando()
    end
    
    def finalDelJuego
      final_juego = false
      i=0
      
      while i<@@NumJugadores && !final_juego
        final_juego = @jugadores.at(i).enBancarrota()
        i = i+1
      end
      
      final_juego
    end
    
    def compareTo(jugador)
      @saldo <=> otro.get_saldo
    end
    
    def ranking # El ranking debe ser público para poder acceder a él desde el controlador
      playersrank = []
      
      jugadores_aux = @jugadores.clone
      
      for i in 0..@@NumJugadores
        max = jugadores_aux.at(0)
        pos = 0
        
        for j in 1..jugadores_aux.size-1
          if max.compare_to(jugadores_aux[j]) < 0
            max = jugadores_aux.at(j)
            pos = j
          end
        end
          
        playersrank.push(max)
        jugadores_aux.delete_at(pos)
        
      end
      
      playersrank  
    end
    
    private # -------------------------------------------------
    
    def avanzaJugador()
      
      # Declaramos al jugador actual y su posicion
      jugadorActual = getJugadorActual()
      posicionActual = jugadorActual.numCasillaActual
      
      # Calculamos su nueva posicion tirando el dado  
      tirada = Dado.instance.tirar()
      posicionNueva = @tablero.nuevaPosicion(posicionActual, tirada)
      
      # Declaramos la casilla en la que está la nueva posición  
      casilla = @tablero.getCasilla(posicionNueva)
      
      # Comprobamos si ha pasado por salida para que el jugador reciba el dinero en tal caso
      contabilizarPasosPorSalida(jugadorActual)
      
      # Movemos al jugador a la nueva posición
      jugadorActual.moverACasilla(posicionNueva)
      
      # Actualizamos la casilla y volvemos a comprobar si ha pasado por salida
      
      casilla.recibeJugador(@indiceJugadorActual, @jugadores)
      contabilizarPasosPorSalida(jugadorActual)
    end
    
    def contabilizarPasosPorSalida(jugadorActual)
      while @tablero.getPorSalida > 0
        jugadorActual.pasarPorSalida
      end
    end
    
    def inicializaTablero(mazo)
      #@tablero = Tablero.new(@@CasillaCarcel)
      
      # Calles
      alquiler = 100
      hipBase = 50
      precioCompra = 150
      precioEdif = 200
      factorRev = 1.2
      cont = 1   # Contador para el nombre de las calles
      
      cantidad_impuesto = 150
      
      # Tablero como en civitasJava
      for i in 1..@@NumCasillas-1 
        case i
          
        when 2 # Sorpresa 1
          @tablero.añadeCasilla(Casilla.newSorpresa(@mazo, "Sorpresa 1"))
          
        when 4 # Impuesto
          @tablero.añadeCasilla(Casilla.newImpuesto( cantidad_impuesto, "Impuesto #{cantidad_impuesto}"))
        
        when 7 # Sorpresa 2
          @tablero.añadeCasilla(Casilla.newSorpresa(@mazo, "Sorpresa 2"))
          
        when 8 # Juez
          @tablero.añadeJuez
          
        when 10 # Sorpresa 3
          @tablero.añadeCasilla(Casilla.newSorpresa(@mazo, "Sorpresa 3"))
          
        when 12 # Parking
          @tablero.añadeCasilla(Casilla.newDescanso("Parking"))
          
        when 3 # Carcel en 3, se añade automaticamente (si no lo ponemos se ejecuta añadir calle en else)
          
        else
          @tablero.añadeCasilla(Casilla.newCalle(TituloPropiedad.new("Calle #{cont}", alquiler, factorRev, hipBase, precioCompra, precioEdif)))
          
        end
        
      end
      
      
    end
    
    def inicializaMazoSorpresas(tablero)
      valor = 200
      ir_casilla_juez = 8
      ir_a_calle = 9
      ir_a_calle2 = 12
      
      # Ir calle 1
      @mazo.alMazo(Sorpresa.newIrCasilla(TipoSorpresa::IRCASILLA, tablero, ir_a_calle, " ¡Vas a la calle de la casilla #{ir_a_calle}!"))
      
      #Juez
      @mazo.alMazo(Sorpresa.newIrCasilla(TipoSorpresa::IRCASILLA, tablero, ir_casilla_juez,  "¡Vas a la calle de la casilla #{ir_casilla_juez}!"))
      
      # Ir calle 2
      @mazo.alMazo(Sorpresa.newIrCasilla(TipoSorpresa::IRCASILLA, tablero, ir_a_calle, " ¡Vas a la calle de la casilla #{ir_a_calle}!"))
      
      # Salir carcel
      @mazo.alMazo(Sorpresa.newEvitaCarcel(TipoSorpresa::SALIRCARCEL, @mazo))
      
      # Ir carcel
      @mazo.alMazo(Sorpresa.newIrCarcel(TipoSorpresa::IRCARCEL, tablero))
      
      # Por jugador, positiva (rebibe) y negativa (paga)
      @mazo.alMazo(Sorpresa.newOtras(TipoSorpresa::PORJUGADOR, valor, " ¡Recibes 200 civiMonedas de cada jugador!"))
      @mazo.alMazo(Sorpresa.newOtras(TipoSorpresa::PORJUGADOR, valor*(-1), " Tienes que pagarle 200 civiMonedas a cada jugador ..."))
      
      # Por casa hotel positiva y negativa
      @mazo.alMazo(Sorpresa.newOtras(TipoSorpresa::PORCASAHOTEL, valor, " ¡Recibes 200 civiMonedas por cada casa y hotel que tengas!"))
      @mazo.alMazo(Sorpresa.newOtras(TipoSorpresa::PORCASAHOTEL, valor*(-1), " Tienes que pagar 200 civiMonedas por cada casa y hotel que tengas ..."))
    
      # Pagar cobrar positiva y negativa
      @mazo.alMazo(Sorpresa.newOtras(TipoSorpresa::PAGARCOBRAR, valor, " ¡Cobras 200 civiMonedas!"))
      @mazo.alMazo(Sorpresa.newOtras(TipoSorpresa::PAGARCOBRAR, valor*(-1), " Tienes que pagar 200 civiMonedas ..."))
      
    end
    
    def pasarTurno
      @indiceJugadorActual = (@indiceJugadorActual+1) % @@NumJugadores
    end
    
  end
end
